### DISCUSSION TITLE
Bugfix & Feature Wishlist

### DISCUSSION THREAD

**Discussion by problame** [zrepl/zrepl#547] [Ideas]
With zrepl 0.5 pretty much done, I'd like to take the opportunity to ask users about which bug fixes and features should be tackled next.
It doesn't matter whether there is already an issue for it.
The purpose of this discussion is to gauge relevance, and also to figure out the big-picture direction for zrepl.
I hope I'll find time to post some ideas myself over the holidays.

**Comment by schulzh** [comment]
I would like some easier way of getting notified of certain error conditions. I know there is the ability to setup Prometheus and Grafana and have it send Notifications when some metric exceeds a threshold, but that is quite a lot of extra effort. I think such notifications should be easy to configure and not require lots of additional software setup.

That being said, the configuration of these notifications and their thresholds should be flexible enough. E.g. I have a replication over a quite unstable connection, so replications frequently fail for various reasons, but overall they replicate fine eventually.

Maybe this could be implemented using the existing Prometheus coding and just building some form of configuring and sending notifications.

Additionally, some more metrics might be needed. I think this [this issue](https://github.com/zrepl/zrepl/issues/353) in particular is quite important, it would be a good basis for notifying about "bigger" replication issues and naturally filter out smaller, unimportant errors.

**Reply by InsanePrawn** [reply]
Any ideas/preferences as to how/where those notifications are sent?
I think we'd rather avoid making specific implementations, e.g. Slack notifications.
[Gotify](https://gotify.net/) comes to mind, but I've never actively used it myself.
Any opinions?

**Reply by schulzh** [reply]
For me just invoking the `mail` command would suffice, but I'm sure others would prefer some more modern approaches. It should be simple though, e.g. for a mail notification, I'd rather not run another service that needs to be configured, monitored, etc.

**Reply by schulzh** [reply]
Maybe just have zrepl run a notification script with a defined set of inputs and include some template/example scripts for sending a mail, calling some webhook, etc.?

**Reply by problame** [reply]
On my setup, I'm using a set of custom icinga2 commands to check the age of the most recent snapshot on the receiver.
The alert threshold is 3h for my server-to-server replications.
And a couple of days for my (manually-triggered) laptop-to-server replications.

That icinga2 setup is older than zrepl itself, and since I tend to run dev versions of zrepl, I like to not rely on it for monitoring.
I use the Grafana dashboard, but mostly to monitor before-after-deployment change in Go runtime metrics.
The only alert I have is the 'number of log messages that require attention' panel.

Now, I think checking the age is the best end-to-end monitoring you can have at the ZFS level.
Even better end-to-end monitoring would be to inject data on the sender and on the receiver, periodically mount the latest snapshot and see if the data is there. But that requires encryption keys on the receiver, and for my laptop-to-server job, that's not going to happen.

Regarding your proposal to make zrepl invoke script / send notifications: I don't like edge-triggered stuff. What if the script fails? Should zrepl retry? How often? How frequently?
I think it's better to have something poll the end-to-end metric, and make alerting decisions based on the poll result. Much more robust in my opinion.

Now, what we could discuss is having the 'most recent snapshot age' metric in Prometheus, or zrepl status for that matter.
Those who don't want a full-blown prometheus + alert-manager/Grafana setup could simply use zrepl status + jq in a cronjob.

**Reply by lapo-luchini** [reply]
As proposed in [a PR 530 comment](https://github.com/zrepl/zrepl/pull/530#discussion_r745419980) I was thinking about a fs-level granularity Prometheus gauge with the timestamp of the last succesfully transmitted snapshot. That would allow interesting graphs and alerts on Prometheus side. I'm not sure I know enough Go to try that, but with a bit of time (or suggestions) I can try. (I would in any case, as I myself want that)

**Reply by Freeaqingme** [reply]
> On my setup, I'm using a set of custom icinga2 commands to check the age of the most recent snapshot on the receiver.

@problame is this something you could share? I found this issue by searching for how to monitor zrepl using Icinga and there doesn't seem to be too much other ideas about this thus far.

**Reply by papamoose** [reply]
Definitely being able to easily send a notification to slack/discord/generic webhook would be nice. I would be happy with the ability to setup a hook post send that would have access to a few environment variables to help determine if a send succeeded.
Obviously, using Prometheus is always going to be preferred in an enterprise environment however a simple method to get notifications would be appreciated if you're not running Prometheus.

My current method involves a python script which queries the Prometheus metrics (`zrepl_replication_last_successful`).

[Discord has a Slack compatibility mode](https://discord.com/developers/docs/resources/webhook#execute-slackcompatible-webhook). [zfs-zed has Slack notification option](https://github.com/openzfs/zfs/blob/f2c5bc150e609a78185ea63c84fce7718f56e28a/cmd/zed/zed.d/zed.rc#L79). Zed's pattern seems reasonable to follow.

Also +1 to gotify support (in addition to, not instead of the above).

**Reply by papamoose** [reply]
[shoutrrr](https://github.com/containrrr/shoutrrr) might also be a possibility.

**Reply by skirmess** [reply]
> I don't like edge-triggered stuff. What if the script fails? Should zrepl retry? How often? How frequently?

I'm using [Uptime Kuma](https://github.com/louislam/uptime-kuma) to monitor my services. With Kuma you can define a push trigger and get URLs like these:

```
https://kuma.example.net/api/push/abcdef123?status=up&msg=OK&ping=
https://kuma.example.net/api/push/abcdef123?status=down&msg=Could%20not%20create%20snapshot&ping=
```

The nice thing is that you can also define a timeout and if the webhook wasn't called for the timeouts duration, it also fails the service.

> could simply use zrepl status + jq in a cronjob.

How would I get the "healthiness" of zrepl from the status?

**Reply by tbleiker** [reply]
@skirmess And where do you call these urls? I had a look "Pre- and Post-Snapshot Hooks" in the docs, but I couldn't figure out how to detect a failure during the job itself in a post-snapshot hook...

**Comment by budachst** [comment]
I'd really like this feature to become available:  introduce conflict_resolution.initial_replication.replicate_all_snapshots #470 

Being able to initially sync a complete ZFS to the destination, or re-sync a long standing zrepl source should be easy to do and not require to have to manually sync the entire zpool beforehand.

And just to stress this a bit more… I just found out that ZFS currently only handles up to 65k Snapshots for  send/recv. Having a ZPOOL with 375 ZFS and 215k Snapshots, there's currently no way of running a simple zfs send -R from the root of the ZPOOL, but you'll have to send each of the 375 ZFS by themselves.

**Reply by problame** [reply]
Thanks for this feedback. Do you have [opinions on syntax](https://github.com/zrepl/zrepl/pull/470#issuecomment-968329346)? The underlying issue is that I'd like to ensure that the syntax we pick is compatible with future conflict resolution stuff, such as, automatic handling of the case where the source side destroys a filesystem and re-creates one with the same name.

**Reply by problame** [reply]
I am finishing @grahamc's work on this in https://github.com/zrepl/zrepl/pull/592 . Upvoters, please consider giving the code a spin before I merge it!

**Comment by khenderick** [comment]
I would like to add #554 to this list.

In summary, I would like to be able to schedule taking a snapshot (and subsequent operations such as replicating). This might be possible by using a newly introduced `cron` type in the `snapshotting` section. It could be considered a more advanced version of the `periodic` type.

My own use-case is using a hook to take a backup of a non-zfs system, downloading it to a zfs dataset and then snapshot/replicate it. I want this to happen at a for the non-zfs system suitable time (off hours).

**Reply by problame** [reply]
Follow-up questions about this:

- Should the snapshot name for all snapshotted filesystems be the time where the `cron` trigger fired? Even if it takes several seconds to snapshot all datasets?
- In your case, "off hours" means probably 'one snapshot per day, at 3 o'clock at night', right?

**Reply by khenderick** [reply]
> Should the snapshot name for all snapshotted filesystems be the time where the cron trigger fired?

What is the current implementation for e.g. the `periodic` time? Assume a user configures the system to take a snapshot every hour, and the first one is taken at `01:00:00` and the snapshotting itself takes 5 seconds, would the time included in the name be `01:00:00` or `01:00:05`? The next snapshot, would it be started at `02:00:00` or at `02:00:05` resulting in a possible time included in the name of `02:00:00`, `02:00:05`, `02:00:10`? I think it would make most sense to use a similar approach for the `cron` type, so it's as consistent as possible across zrepl.

For my own use-case where I'd use a hook to backup a non-zfs system, the time for taking the backup itself, transferring it to the zfs dataset, and taking the actual backup itself would introduce time offsets anyway. I think it would be hard to get the timestamp in the snapshot name correctly represent the timestamp of the data inside.

> In your case, "off hours" means probably 'one snapshot per day, at 3 o'clock at night', right?

Yes, for my own use-case something like an exact time would be sufficient (e.g. being able to configure to take a snapshots every day at 3 o'clock at night), but as you mentioned `cron` in one of the comments in #397 I thought this might allow more flexible setups (e.g. only in the weekend).

**Reply by lapo-luchini** [reply]
Myself I'd really love if snapshots were taken at the same time on all filesystems, i.e. using recursive snapshots o having multiple fs parameters in the same `zfs snap` line, so that they all have the same exact name and, more importantly, that a system-wide consistent snapshot can be taken.

Right now I'm doing that with a separate recursive snapshot cron (and that snap gets transferred with the ones created by zrepl) but would love to drop the external snapshot tool and have the same snapshot being used for consistent snapshot and for zrepl at the same time.

**Reply by problame** [reply]
@khenderick it's counting the interval since the last trigger point, but doesn't snapshot until the previous snapshot is done. So, if it triggers at time X, and snapshotting takes Y seconds, it would trigger at `MAX(X+Y, X+$periodic)`. Further, there is a sync-up phase on daemon startup. Find the logic here:

sync-up:

https://github.com/zrepl/zrepl/blob/b2c6e51a43c70d0b06690cff3675aab4d365301a/daemon/snapper/snapper.go#L191-L216

general wait time:

https://github.com/zrepl/zrepl/blob/b2c6e51a43c70d0b06690cff3675aab4d365301a/daemon/snapper/snapper.go#L359-L384

**Reply by problame** [reply]
@lapo-luchini There is a long-standing issue to use ZFS channel programs for consistent snapshots ( #46 ). Basically we'd need code that compiles a zrepl `filesystems` filter to a channel program. The only weak spot with channel programs is that they're only atomic, but not transactional. I.e., if we run over the amount of allowed lua instructions / mem limit, only some snapshots will have been taken. We'd need to detect that error condition in zrepl and then clean-up, and potentially retry with a higher limit.

The alternative is to do a recursive snapshot and then delete snaps that don't match `filesystems`, but I don't like that approach.

Perhaps this should be a separate wishlist item, so it doesn't get lost here?

**Reply by mdlayher** [reply]
I don't have a need for scheduling snapshots, but I do have a need to schedule replication. NixOS ships with some hooks to set up `zfs-auto-snapshot` pretty easily, so I complemented that with a quick systemd unit and timer to trigger replication:

```
[matt@servnerr-3:~]$ systemctl cat zrepl-signal-jobs.timer
# /etc/systemd/system/zrepl-signal-jobs.timer
[Unit]
PartOf=zrepl-signal-jobs.service

[Timer]
OnCalendar=hourly
Unit=zrepl-signal-jobs.service

[matt@servnerr-3:~]$ systemctl cat zrepl-signal-jobs.service
# /etc/systemd/system/zrepl-signal-jobs.service
[Unit]

[Service]
Environment="LOCALE_ARCHIVE=/nix/store/vxxqn7g4ds1vlhdcbpi1m2hw4xqm80r9-glibc-locales-2.33-78/lib/locale/locale-archive"
Environment="PATH=/nix/store/yyr35ka8sqq6mym9r20djvsjl8rzg9ah-zrepl-0.5.0/bin:/nix/store/5wch96kji9zlffxjqpjdrszjzp4i7m3a-coreutils-9.0/bin:/nix/store/0hj8idivq7434gbsi0jx3dfhd7bsgb3a-findutils->
Environment="TZDIR=/nix/store/6cc9lmldp0w0vvgr5k1h1gr8i6hn791b-tzdata-2021c/share/zoneinfo"

ExecStart=/nix/store/livfw3wv30jf3azv3wvrxx6gwgwr85np-unit-script-zrepl-signal-jobs-start/bin/zrepl-signal-jobs-start 
Type=oneshot

[matt@servnerr-3:~]$ cat /nix/store/livfw3wv30jf3azv3wvrxx6gwgwr85np-unit-script-zrepl-signal-jobs-start/bin/zrepl-signal-jobs-start
#!/nix/store/kq5si67lhbidwkw444kc3lz3x3l7nxy6-bash-5.1-p8/bin/bash -e
zrepl signal wakeup primary_to_secondary
zrepl signal wakeup primary_to_backup0
zrepl signal wakeup primary_to_backup1
```

This does the job and my `zfs-auto-snapshot` snapshots all get replicated hourly. I was curious if there'd be a way to trigger a periodic replication _without prior snapshotting_ in the config, but have no strong feelings about whether or not that should be included since I am making use of two different systems here.

**Reply by problame** [reply]
>  if there'd be a way to trigger a periodic replication without prior snapshotting in the config, 

There's a WIP PR for this, but I preferred a different design. Started implementation but never finished.
https://github.com/zrepl/zrepl/pull/462

@mdlayher could you turn this into an independent entry on this wishlist?

**Reply by mdlayher** [reply]
Can do!

**Reply by spi43984** [reply]
> I would like to add #554 to this list.
> 
> In summary, I would like to be able to schedule taking a snapshot (and subsequent operations such as replicating). This might be possible by using a newly introduced `cron` type in the `snapshotting` section. It could be considered a more advanced version of the `periodic` type.
> 
> My own use-case is using a hook to take a backup of a non-zfs system, downloading it to a zfs dataset and then snapshot/replicate it. I want this to happen at a for the non-zfs system suitable time (off hours).

I second this idea. I also would like to create snapshots at a particular time and run the replication at off hours rather then at periodic time intervals.

Something else I would like to add to the list is some kind of benchmark support to see the throughput (like for different servers and different transports, as e.g. scp is doing). Sorry if there is something and I've just missed it but the only thing I could find is some live view in `zrepl status` which isn't of much help. I'd like to see how much data was transmitted in what amount of time for each filesystem and in total.

**Comment by TheCasualObserver** [comment]
I would kindly ask you to have a look at the bug report #490 for the next release.

**Reply by lapo-luchini** [reply]
I have a similar problem and would be interested in this as well, though different, but I didn't examine the exact problem so far; my "chained" use-case is A→B pushed on A side then B→C pulled from C side, doesn't conflict and usually work, but rarely kills the server on B, which at the moment I worked around with a software watchdog.

**Reply by problame** [reply]
> , but rarely kills the server on B, which at the moment I worked around with a software watchdog.

@lapo-luchini  please collect the stack trace, and core dump, when it crashes the next time (`GOTRACEBACK=crash`, learn more [here](https://www.jetbrains.com/help/go/exploring-go-core-dumps.html#view-go-core-dumps-in-goland))

**Reply by lpulley** [reply]
Is this "push to B, pull from B" a recommended way to handle A>B>C replication? I'm trying to figure out how to do something just like that.

**Reply by problame** [reply]
Push vs pull doesn't matter as long as all jobs involved have disjoint names, and you don't need fan-out (for fan-out, you'll need something like https://github.com/zrepl/zrepl/issues/380#issuecomment-1273843713 )

**Reply by lpulley** [reply]
I decided A>B>C isn't ideal for me because I still want A>C if B is down. I suppose A>B>C isn't *exactly* right, since my A and B are on the same machine, with C remote. I'm thinking of setting up fan-out with systemd timer wakeups for the push jobs -- unless you think `source`/`pull` is more stable than `push`/`sink` for fan-out?

Also, are there any changes in 0.6.0 that should affect fan-out?

**Reply by problame** [reply]
There are no changes that affect fan-out in 0.6

It works with push jobs because each push job has a different name and can hence maintain different replication cursors.
To make it work with pull jobs, you need a separate source job per pull job so that the replication cursor bookmark names don’t conflict.

In general, I would advise to diversify the backup strategy instead of going all in on zfs send/recv.

**Reply by lpulley** [reply]
The plan is to have

- [zrepl] push A to local B (keeping a fading history)
- [zrepl] push A to remote C (keeping a fading history)
- [restic] backup A to remote D (keeping a fading history)

So I do have a non-ZFS (file-based) backup included in this plan. Don't worry, I don't trust ZFS *completely*. 😆

**Comment by eoli3n** [comment]
Ability to disable jobs : https://github.com/zrepl/zrepl/issues/333

**Reply by problame** [reply]
@eoli3n I [responded there](https://github.com/zrepl/zrepl/issues/333#issuecomment-1030595099), and I'm interested in your thoughts. Please comment there.

**Comment by papamoose** [comment]
Ability to specify the destination dataset per source dataset.

e.g.
tank/foo => tank/foo
tank/bar => tank/containers/bar

Additionally if and some other value you set to enable this:
```
  filesystems: {
    "tank<": true,
  }
```
On the receiving side the file systems would be setup identically.
e.g. both sides would look like this:
`tank/foo`
`tank/bar/baz`

**Reply by mdlayher** [reply]
I ran into the same thought when setting up zrepl today. I am backing up locally from one zpool to a hot backup pool and a cold backup pool that is a single disk.

I had assumed if I have `primary/media`, I'd get `secondary/media`, but the current hierachy is `secondary/local/primary/media`. Not really a huge deal I suppose (and I was able to rename a large existing dataset to match), but in my case I wouldn't ever anticipate receiving pool streams from multiple senders since I'm just doing a local copy.

**Reply by problame** [reply]
@papamoose I understand you want the ability to map arbitrarily between sender and receiver filesystem names.
Whereas for @mdlayher, it would probably be sufficient to support `root_fs=""`.

For both of you, the issue that's worth subscribing to is probably https://github.com/zrepl/zrepl/issues/253.

A few follow-up questions to @mdlayher :

* So, you are pushing a set of filesystems from `primary` to `secondary/local`. And then, from `secondary/local`, to yet another cold backup pool?
*  Regardless of the previous question: would you want `primary`'s content to be replicated to `secondary` as well, then? Or are you not interested in replicating `primary`, but only its children?
(Because: `primary` is not just the pool name, it's also a regular dataset, that could contain user data).

**Reply by mdlayher** [reply]
Thanks for following up! Here is my current generated configuration for reference: https://gist.github.com/mdlayher/36c201634be0e2b9593f266e0eb777ec

> So, you are pushing a set of filesystems from primary to secondary/local. And then, from secondary/local, to yet another cold backup pool?

I would like to replicate `primary` to `secondary`, but since I only have a single client (the local host) I just set `client_identity = local`, so yep the computed path ends up being `primary/media` to `secondary/local/primary/media`.

I have two single disk pools called `backup0` and `backup1` and replicate directly from `primary` to those when the pools are imported and online. These are cold storage disks that I will plug in, replicate, and then probably take offsite or store in a safe.

> Regardless of the previous question: would you want primary's content to be replicated to secondary as well, then? Or are you not interested in replicating primary, but only its children?
> (Because: primary is not just the pool name, it's also a regular dataset, that could contain user data).

I'm only interested in replicating the children, yes! I've generally been treating non-leaf datasets as just containers for other datasets.

After thinking on this for a couple of days, it is a bit nice to have the `secondary/local` prefix because I can only set ZFS encryption on a child dataset rather than the entire zpool. This means that I can guarantee everything written in under that child is encrypted, and if I understand correctly, should also simplify unlocking the `backup{0,1}` pools using ZFS native encryption.

---

So really there are no necessary changes here since no software I am using demands a particular layout in my backup pools. I just wanted to share my initial thoughts.

**Reply by problame** [reply]
Yes, encryption is another problem child wrt `root_fs=""`. I think you can set it on the root dataset of a pool, but you can't `recv -F` then.

Regarding encryption on the encrypt-on-recv: make sure to use zrepl 0.5. see changelog for details.

**Reply by mdlayher** [reply]
NixOS 21.11 installed 0.4 by default so I learned that one the hard way unfortunately, but I can confirm that 0.5 worked great! Thanks.

**Comment by mdlayher** [comment]
As a follow up to https://github.com/zrepl/zrepl/discussions/547#discussioncomment-2151714 and as requested by @problame, I would like a way to be able to trigger replication on a schedule without necessarily creating a snapshot first.

I have a workaround that works, but frankly, I get the feeling I should just embrace the zrepl periodic snapshotting mechanism rather than trying to make `zfs-auto-snapshot` (as configured by NixOS) play nicely.

But perhaps it'd be nice to just have some way to say "run this replication job every hour" or similar, since you could pretty easily set up some Go tickers to send a signal at regular intervals.

**Reply by mdlayher** [reply]
I switched over to zrepl snapshotting today and have disabled `zfs-auto-snapshot`, so I think my request is now obsolete!

**Comment by eoli3n** [comment]
https://github.com/zrepl/zrepl/issues/577 : introduce '>' value for sub datasets only in filesystems key
https://github.com/zrepl/zrepl/issues/579 : store bucket value for snapshots in a property

**Comment by sabirovrinat85** [comment]
Is there anything like that [zfs-check](https://github.com/psy0rz/zfs_autobackup/wiki/zfs-check) ? If no, then maybe it's worth considering as an additional feature
